include/claim_strategy.hpp

```diff
@@ -0,0 +1,300 @@
+#ifndef GNURADIO_CLAIM_STRATEGY_HPP
+#define GNURADIO_CLAIM_STRATEGY_HPP
+
+#include <concepts>
+#include <cstdint>
+#include <memory>
+#include <span>
+#include <stdexcept>
+#include <vector>
+
+#ifndef GNURADIO_SEQUENCE_HPP
+#include <sequence.hpp>
+#endif
+#ifndef GNURADIO_WAIT_STRATEGY_HPP
+#include <wait_strategy.hpp>
+#endif
+
+namespace gr {
+
+struct NoCapacityException : public std::runtime_error {
+    NoCapacityException() : std::runtime_error("NoCapacityException"){};
+};
+
+// clang-format off
+
+template<typename T>
+inline constexpr bool isClaimStrategy = requires(T /*const*/ t, const std::vector<std::shared_ptr<Sequence>> &dependents, const int requiredCapacity,
+        const std::int64_t cursorValue, const std::int64_t sequence, const std::int64_t availableSequence, const std::int32_t n_slots_to_claim) {
+    { t.hasAvailableCapacity(dependents, requiredCapacity, cursorValue) } -> std::same_as<bool>;
+    { t.next(dependents, n_slots_to_claim) } -> std::same_as<std::int64_t>;
+    { t.tryNext(dependents, n_slots_to_claim) } -> std::same_as<std::int64_t>;
+    { t.getRemainingCapacity(dependents) } -> std::same_as<std::int64_t>;
+    { t.publish(sequence) } -> std::same_as<void>;
+    { t.isAvailable(sequence) } -> std::same_as<bool>;
+    { t.getHighestPublishedSequence(sequence, availableSequence) } -> std::same_as<std::int64_t>;
+};
+
+template<typename T>
+concept ClaimStrategy = isClaimStrategy<T>;
+
+namespace claim_strategy::util {
+constexpr bool isPower2(std::unsigned_integral auto value) { return !(value == 0) && !(value & (value - 1)); }
```

Comment:
std::has_single_bit?

Reply:
tried this but unfortunately doesn't evaluate as a `constexpr` in a template-requires environment.